Update README for release 3.1.0 (#324)
microsoft/FeatureManagement-Dotnet: README.md

In the above example, `FeatureW` specifies a `RequirementType` of `All`, meaning all of it's filters must evaluate to true for the feature to be enabled. In this case, the feature will be enabled for 50% of users during the specified time window.
    
## Service Registration

> Feature flags rely on .NET Core dependency injection. We can register the feature management services using standard conventions. By default, the feature manager and feature filters should be registered as singleton services by calling `AddFeatureManagement`.

Reviewer: I'd leave out the newly added text to keep the simple case simple.

I'd suggest any new content or clarifications be a part of the new section(s).

Author: Sounds good.

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddFeatureManagement();
    }
}
```

By default, the feature manager retrieves feature flag configuration from the "FeatureManagement" section of the .NET Core configuration data. If the "FeatureManagement" section does not exist, the configuration will be considered empty.

> You can also specify that feature flag configuration should be retrieved from a different configuration section by calling configuration.GetSection and passing in the name of the desired section. The following example tells the feature manager to read from a different section called "MyFeatureFlags" instead:

Reviewer: ```suggestion
You can also specify that feature flag configuration should be retrieved from a different configuration section by the section to `AddFeatureManagement`. The following example tells the feature manager to read from a different section called "MyFeatureFlags" instead:
```

> ``` C#

Reviewer: can you put a newline before the code snippet.

services.AddFeatureManagement(configuration.GetSection("MyFeatureFlags"));
```

### Scoped Feature Management Services
By default, the feature manager and feature filters should be registered as singleton services by calling `AddFeatureManagement`.

> There are scenarios that feature filters are not necessarily be singleton. For example, users may want to use feature filters which consume scoped services for context information. In this case, the feature manager and feature filters should be registered as scoped services by calling `AddScopedFeatureManagement`:

Reviewer: ```suggestion
The `AddFeatureManagement` method adds feature management services as singletons within the application, but there are scenarios where it may be necessary for feature management services to be added as scoped services instead. For example, users may want to use feature filters which consume scoped services for context information. In this case, the feature manager and feature filters should be registered as scoped services by calling `AddScopedFeatureManagement`:
```

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddFeatureManagement();
    }
}
```

By default, the feature manager retrieves feature flag configuration from the "FeatureManagement" section of the .NET Core configuration data. If the "FeatureManagement" section does not exist, the configuration will be considered empty.

You can also specify that feature flag configuration should be retrieved from a different configuration section by passing the section to `AddFeatureManagement`. The following example tells the feature manager to read from a different section called "MyFeatureFlags" instead:

``` C#
services.AddFeatureManagement(configuration.GetSection("MyFeatureFlags"));
```

### Scoped Feature Management Services

> By default, the feature manager and feature filters should be registered as singleton services by calling `AddFeatureManagement`.

Reviewer: With the update to the following paragraph, this sentence is not needed.

{
  public void ConfigureServices(IServiceCollection services)
  {
      services.AddFeatureManagement()
              .AddFeatureFilter<MyCriteriaFilter>();
  }
}
```

> This tells the feature manager to use the "FeatureManagement" section from the configuration for feature flag settings. It also registers a customized feature filter `MyCriteriaFilter`. Please refer to the `Implementing a Feature Filter` section below for more details.

Reviewer: Why bring up filters here? It's a bit early.

Author: I mentioned that "By default, the feature manager and feature filters should be registered as singleton services by calling AddFeatureManagement." So I bring up feature filter here.

Author: The concept of feature filter has been introduced in the `Feature Filters` section. Would it be fine to bring up filters here?

You can also specify that feature management configuration should be retrieved from a different configuration section by calling configuration.GetSection and passing in the name of the desired section. The following example tells the feature manager to read from a different section called "MyFeatureFlags" instead:
``` C#
…
services.AddFeatureManagement(configuration.GetSection("MyFeatureFlags"));
…
```

> **Advanced:** By default, the feature manager retrieves feature flag configuration from the "FeatureManagement" section of the .NET Core configuration data. If the "FeatureManagement" section does not exist, the configuration will be considered empty. When a specific configuration section is provided to `AddFeatureManagement()`, the feature manager will fall back to the root of that configuration section to look for feature definitions, if there is no "FeatureManagement" section found. 

Reviewer: > When a specific configuration section is provided to `AddFeatureManagement()`, the feature manager will fall back to the root of that configuration section to look for feature definitions, if there is no "FeatureManagement" section found.

I think we can simplify this to

"When a specific configuration section is passed when registering feature management, the "FeatureManagement" section is not needed."

Reviewer: Actually, I might suggest that we move this "Advanced" note to be directly under the section detailing the plain "AddFeatureManagement", and then move on to describing the alternative approach which lets you pass a configuration section directly. In the alternative approach I think it should be clear enough by our explanation that the section is where the flags are defined.

### Scoped Feature Management Services

> There are scenarios that feature filters are not necessarily be singleton ([#15](https://github.com/microsoft/FeatureManagement-Dotnet/issues/15), [#258](https://github.com/microsoft/FeatureManagement-Dotnet/issues/258)). For example, users may want to use feature filters which consume scoped services for context information. In this case, the feature management should be registered as scoped by calling `AddScopedFeatureManagement`.

Reviewer: I'm not fond of the idea of linking to PRs/Issues in documentation.

This example demonstrates how to register feature manager and feature filters as scoped services to allow the feature filter `MyCriteriaFilter` to consume the scoped context provider service `MyContextProvider`.
``` C#
public class MyCriteriaFilter : IFeatureFilter
{
    private readonly IContextProvider _contextProvider;

    public MyCriteriaFilter(IContextProvider contextProvider)
    {
        _contextProvider = contextProvider;
    }
}
```

``` C#
…
services.AddScopedFeatureManagement()

>         .AddFeatureFilter<MyCriteriaFilter>();

Reviewer: I'd leave out the filter. This is kind of an "all-else-follows" section. Just mention that feature manager, as well as all registered filters are added as scoped.

Author: Here I want to present an example which demonstrates how to register feature manager and feature filters as scoped services to allow the feature filter to consume the scoped context provider service

Are you suggesting that I should remove this example?

Creating a feature filter provides a way to enable features based on criteria that you define. To implement a feature filter, the `IFeatureFilter` interface must be implemented. `IFeatureFilter` has a single method named `EvaluateAsync`. When a feature specifies that it can be enabled for a feature filter, the `EvaluateAsync` method is called. If `EvaluateAsync` returns `true` it means the feature should be enabled.

The following snippet demonstrates how to add customized feature filters `MyFirstCriteriaFilter` and `MySecondCriteriaFilter`.

``` C#
services.AddFeatureManagement()

>         .AddFeatureFilter<MyFirstCriteriaFilter>()

Reviewer: I don't see the benefit of adding a new filter in this code snippet.

Author: We mentioned "Feature filters are registered by calling `AddFeatureFilter<T>` on the `IFeatureManagementBuilder` returned from `AddFeatureManagement`."
But we do not mention that AddFeatureFilter<T> will also return `IFeatureManagementBuilder`. So I register multiple filters in the example here.

Now, I like the idea to keep the simplicity. So I decided to remove it.

        .AddFeatureFilter<MySecondCriteriaFilter>();
```

Feature filters are registered by calling `AddFeatureFilter<T>` on the `IFeatureManagementBuilder` returned from `AddFeatureManagement`. These feature filters have access to the services that exist within the service collection that was used to add feature flags. Dependency injection can be used to retrieve these services.

**Note:** When filters are referenced in feature flag settings (e.g. appsettings.json), the _Filter_ part of the type name should be omitted. Please refer to the `Filter Alias Attribute` section below for more details.

> 

Reviewer: This is already clear from the description in the AddScopedFeatureManagement section.

Author: Removed.

Creating a feature filter provides a way to enable features based on criteria that you define. To implement a feature filter, the `IFeatureFilter` interface must be implemented. `IFeatureFilter` has a single method named `EvaluateAsync`. When a feature specifies that it can be enabled for a feature filter, the `EvaluateAsync` method is called. If `EvaluateAsync` returns `true` it means the feature should be enabled.

The following snippet demonstrates how to add customized feature filters `MyFirstCriteriaFilter` and `MySecondCriteriaFilter`.

``` C#
services.AddFeatureManagement()
        .AddFeatureFilter<MyFirstCriteriaFilter>()
        .AddFeatureFilter<MySecondCriteriaFilter>();
```

> Feature filters are registered by calling `AddFeatureFilter<T>` on the `IFeatureManagementBuilder` returned from `AddFeatureManagement`/`AddScopedFeatureManagement`. The feature filters will be registered as the same type as the feature manager. For instance, if the feature manager is registered as a singleton, feature filters will also be registered as singletons. Feature filters have access to the services that exist within the service collection that was used to add feature flags. Dependency injection can be used to retrieve these services.

Reviewer: Suggestion: Keep the paragraph as it is. Add a new paragraph that describes the consequences of DI scope.

### Built-In Feature Filters

> There a few feature filters that come with the `Microsoft.FeatureManagement` package: `PercentageFilter`, `TimeWindowFilter`, `ContextualTargetingFilter` and `TargetingFilter`. All filters, except for the `TargetingFilter`, are added automatically when calling `AddFeatureManagement`/`AddScopedFeatureManagement`. The `TargetingFilter` is added with the `WithTargeting` method that is detailed in the `Targeting` section below.

Reviewer: This change seems unnecessary to me.

The implementation type used for the `ITargetingContextAccessor` service must be implemented by the application that is using the targeting filter. Here is an example setting up feature management in a web application to use the `TargetingFilter` with an implementation of `ITargetingContextAccessor` called `HttpContextTargetingContextAccessor`.

``` C#
services.AddFeatureManagement()
        .WithTargeting<HttpContextTargetingContextAccessor>();
```

> 

Reviewer: > The targeting context accessor and targeting filter will be registered as the same type as the feature manager.

I'd like this topic to stick in the single section discussing service lifetime. As long as we say there a statement that all filters are added with the same service lifetime then we ashould be covered.

The implementation type used for the `ITargetingContextAccessor` service must be implemented by the application that is using the targeting filter. Here is an example setting up feature management in a web application to use the `TargetingFilter` with an implementation of `ITargetingContextAccessor` called `HttpContextTargetingContextAccessor`.

``` C#
services.AddFeatureManagement();

>         .AddFeatureFilter<BrowserFilter>()

Reviewer: Why add the BrowserFilter in the targeting example?

Author: I want to make it more clear how to call the `WithTargeting<>()` method. We didn't mention `AddFeatureFilter` method will return `IFeatureManagementBuilder`.

Author: Removed.
